### DISCUSSION TITLE
The v4 programming model for Node.js is now Generally Available! 🎉

### DISCUSSION THREAD

**Discussion by ejizba** [Azure/azure-functions-nodejs-library#152] [Announcements]
The v4 model was designed to give JavaScript and TypeScript developers a more flexible & intuitive experience when authoring apps on Azure Functions. Read our [blog post](https://techcommunity.microsoft.com/t5/apps-on-azure-blog/azure-functions-node-js-v4-programming-model-is-generally/ba-p/3929217) for more information.

Please use this discussion thread to let us know what you think of the new model, good or bad! We can't wait to make it even better!

**Comment by Krucial92** [comment]
Nice work @ejizba! Loving all of the changes made. Does this mean we can now remove the `AzureWebJobsFeatureFlags`: `EnableWorkerIndexing` config setting in the function apps & local environment too?

**Reply by ejizba** [reply]
Yep! If running locally, you'll need core tools v4.0.5382+ which was just released this week. If running in Azure, you'll want host v4.25+ which finished rolling out last week.

**Comment by thammons** [comment]
v3 was painful.
 v4 is comparatively a dream!

When will Azure Static Webapps (with managed functions) be able to support host v4.25?

**Reply by ejizba** [reply]
I believe this should be working already for static web apps. If you're seeing issues, please try sharing your app name so that we can investigate. Instructions to do so can be found here: https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately

**Reply by thammons** [reply]
Took me a while to find it.. but @azure/functions is no longer a devdependancy... and it's hosted!

Dig through the [migration](https://learn.microsoft.com/en-us/azure/azure-functions/functions-node-upgrade-v4?tabs=v4) guide closely friends!

**Comment by Yvand** [comment]
I migrated my project to the v4 programming model and it works great.
But we seem to lose the ability to quickly enable/disable a function from the Azure portal UI, and now the portal gives a wrong information on its status. It merely creates and reads the app setting `AzureWebJobs.<FUNCTION_NAME>.Disabled`, that no longer has any effect.

Am I correct?
If so, what is the best way to replicate this feature?
I created a custom app setting "functionName_disabled":
- If I read it before registering the function, it works but the function disappears from the portal (which makes sense).
- If I read it at the start of the function's handler, then the portal reports that the function is enabled, even if it does nothing.

Is there a better solution?
Do you plan to re-implement this?

**Reply by ejizba** [reply]
You can use a custom setting and it's a totally valid way to do it in the new model, but as you said it has downsides. The app setting `AzureWebJobs.<FUNCTION_NAME>.Disabled` is the recommended way to disable functions, and it should still work. I just tried today and it was working for me. Can you share your app name and a rough time period when the disabled setting wasn't working? Instructions here: https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately

**Reply by Yvand** [reply]
@ejizba thank you for your reply. I tried again and I confirm that the app setting `AzureWebJobs.<FUNCTION_NAME>.Disabled` no longer works after I migrated to the model v4.
The service runs in the region `West Europe`. Here is the relevant log when the function was successfully called, despite it appears disabled in the Azure portal:

```
2023-10-04 07:49:01.248 Executing 'Functions.HelloHttp' (Reason='This function was programmatically called via the host APIs.', Id=1dcd5f09-371a-4eb4-b926-d19636bc16b4)
2023-10-04 07:49:01.336 Executed 'Functions.HelloHttp' (Succeeded, Id=1dcd5f09-371a-4eb4-b926-d19636bc16b4, Duration=91ms)
```

**Reply by ejizba** [reply]
Thanks for the info. How are you executing the function? I didn't realize, but it seems like disabled functions can always be called by administrators: https://learn.microsoft.com/en-us/azure/azure-functions/disable-function?tabs=portal#run-a-disabled-function

Fyi some tools (like "Code + Test" in the portal) probably use those admin permissions to execution functions. If you try to make a request to the url directly without using any azure-specific tooling, does it work or does it return 404?

**Reply by Yvand** [reply]
I didn't know this, very nice catch.
I call the HTTP function using Postman, but I always pass the master key since I set the authLevel to 'function'.
I tested using another key, and I confirm I get a HTTP 404 if the function is disabled.
Thank you for your help

**Comment by jvirtala** [comment]
I think the new model is quite easy to approach and the added type safety is a big plus.

I did run into some magic when I was converting one of our functions.
Consider a case where I have a HTTP function and I want to add an `extraOutput` of `storageBlob`:
```typescript
const blobOutput = output.storageBlob({
  path: "container/{id}",
  connection: "storagekey",
});

export async function worker(
  request: HttpRequest,
  context: InvocationContext
): Promise<HttpResponseInit> {
  const jobId = request.params["id"];
  const jsonBody = await request.json();
  context.extraOutputs.set(blobOutput, jsonBody);
  return {body: "ok"};
}

app.http("worker", {
  methods: ["POST"],
  route: "worker/{id}",
  extraOutputs: [blobOutput],
  handler: worker,
});
``` 

It's unclear to me how I can modify or set the variable `id` of `storageBlob`. Here it picks it up from the params of the request, I presume, but could I set it myself in the `http` function?

**Reply by jvirtala** [reply]
Actually, immediately as I posted this I got it that it's the model I pass on to `context.extraOutputs.set(blobOutput, jsonBody);` but the confusion started because I got different ID than I expected. I have `id` defined in both, route parameter and body, so it seems the `blobOutput` defaults to route parameter.

**Reply by ejizba** [reply]
Hi @janiilmari, glad you got it figured out! Here are some docs that might be relevant to your situation:
- https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?pivots=programming-language-javascript#using-route-parameters
- https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-expressions-patterns

If you want the most control in terms of setting the "id" variable, your best bet is to use azure sdks [`@azure/storage-blob`](https://www.npmjs.com/package/@azure/storage-blob) directly. If you want to use an output binding, your choices are limited to the options described in those docs.

Regardless, I created a new GitHub issue to track improvements to this area - if you have any suggestions please add a comment there: https://github.com/Azure/azure-functions-nodejs-library/issues/167

**Comment by sandeepsuresh16** [comment]
Hi @ejizba, I am working with Azure Functions that use SignalR bindings. Can you suggest some articles or documents related to the v4 programming model for the signalR bindings?

**Reply by ejizba** [reply]
Hi @sandeepsuresh16 if you want to use SignalR, you'll need to use a generic binding as shown [here](https://learn.microsoft.com/en-us/azure/azure-functions/functions-reference-node?tabs=javascript%2Clinux%2Cazure-cli&pivots=nodejs-model-v4#generic-inputs-and-outputs). If you would like more "official" support for SignalR in the v4 model, please [file a new issue](https://github.com/Azure/azure-functions-nodejs-library/issues/new). We add binding support based on usage and user feedback

**Reply by sandeepsuresh16** [reply]
Hi @ejizba, thank you for the reply. Could you please provide a sample code snippet for the following binding in V4?
![binding_v3](https://github.com/Azure/azure-functions-nodejs-library/assets/69909315/e1ea3512-ad70-4920-8e4d-88e785cc0753)

**Reply by ejizba** [reply]
## JavaScript

```javascript
const { app, output } = require('@azure/functions');

const signalROutput = output.generic({
    type: 'signalR',
    hubName: 'devstackhub',
    connectionStringSetting: 'AzureSignalRConnectionString'
});

app.http('httpTrigger1', {
    methods: ['GET', 'POST'],
    authLevel: 'function',
    extraOutputs: [signalROutput],
    handler: async (request, context) => {
        context.log(`Http function processed request for url "${request.url}"`);

        const name = request.query.get('name') || await request.text() || 'world';

        context.extraOutputs.set(signalROutput, { example: 'value' });

        return { body: `Hello, ${name}!` };
    }
});
```

## TypeScript

```typescript
import { app, HttpRequest, HttpResponseInit, InvocationContext, output } from "@azure/functions";

const signalROutput = output.generic({
    type: 'signalR',
    hubName: 'devstackhub',
    connectionStringSetting: 'AzureSignalRConnectionString'
});

export async function httpTrigger1(request: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> {
    context.log(`Http function processed request for url "${request.url}"`);

    const name = request.query.get('name') || await request.text() || 'world';

    context.extraOutputs.set(signalROutput, { example: 'value' });

    return { body: `Hello, ${name}!` };
};

app.http('httpTrigger1', {
    methods: ['GET', 'POST'],
    authLevel: 'function',
    extraOutputs: [signalROutput],
    handler: httpTrigger1
});
```

**Comment by yuramag** [comment]
Hi, can you please help me to setup a call to create a v4 Function App using Resource Manager API (`Azure.ResourceManager.AppService`)? I even created a bug describing the problem here: https://github.com/Azure/azure-sdk-for-net/issues/37919. 

Basically I am trying to deploy a Function App on the fly, per user request, using `Azure.ResourceManager.AppService` API. It used to work for v3, but when I deploy using the same approach to v4 (of course using new folder structure), it kind of succeeds, but I get a blank screen in "Code + Test" tab in the Azure portal (no errors displayed either), and HTTP 404 when I call the function via HTTP.

**Reply by ejizba** [reply]
Hi @yuramag, I see how you set function.json, but I don't see how you set index.js. How do you set index.js for model v3? In general, you should think of all the files in your app as a unit and deploy all at once instead of using `CreateOrUpdateAsync` to edit a single function at a time. You should theoretically still be able to do this "on the fly", but I don't exactly understand your scenario.

Make sure you take a look at the [migration guide](https://learn.microsoft.com/en-us/azure/azure-functions/functions-node-upgrade-v4?tabs=v4). There are some steps that you probably haven't done in your v4 testing which are causing the problems. For example, you need a package.json file that has a "main" field and lists `@azure/functions` as a dependency. You also need to run `npm install` to install that dependency. AFAIK the `CreateOrUpdateAsync` api _only_ creates the "function.json" file, so it does nothing to help for model v4.

**Reply by yuramag** [reply]
Thanks for the information, it was very helpful! Turns out it was missing the `npm install`. When I run it manually after deploying I see that `node_modules` shows up in the folder structure and the app starts to work as expected.

Is it possible to trigger this command programmatically from code?

**Reply by yuramag** [reply]
BTW, I forgot to mention that I removed the `app.GetSiteFunctions().CreateOrUpdateAsync()` call entirely (thanks for the hint). This function creates an unnecessary folder at root (with `function.json`) and prevents the app from functioning.

**Reply by ejizba** [reply]
You can set your app up to use [remote build](https://learn.microsoft.com/en-us/azure/azure-functions/functions-deployment-technologies?tabs=windows) after every time you deploy. Remote build is more complicated to use on some os/sku combinations than others fyi. But even in that case you still have to do a deploy, so I would personally prefer to just run `npm install` myself before deploying.

What OS and sku (consumption/premium/etc.) are you on? I would be careful about manually running `npm install` _after_ you deploy. If your app recycles instances (which can often happen depending on your sku), you might lose those node_modules.

**Comment by Crisfole** [comment]
It doesn't look like you've released Docker images for the newest core tools and function runtimes? (I'm looking at the release announcement:

> You must be on runtime v4.25+ in Azure or core tools v4.0.5382+ if running locally to benefit from this change.

But when running locally with `mcr.microsoft.com/azure-functions/node:4-node18-core-tools` (I have done a docker pull to make sure I'm up to date) I see the following on startup:

```
my-app-name-functions-1  | Azure Functions Core Tools
my-app-name-functions-1  | Core Tools Version:       4.0.5198 Commit hash: N/A  (64-bit)
my-app-name-functions-1  | Function Runtime Version: 4.21.1.20667
```

It was working beautifully locally on the beta, but now it just doesn't find any job functions (Presumably because I removed the flag and tried updating the functions runtime). Any advice?

**Reply by ejizba** [reply]
Not sure why the images are so out of date, I'll have to follow up with another team and get back to you. Since the host version is that old, you'll have to keep the flag for now

**Reply by Crisfole** [reply]
No worries. I just copied/pasted the Dockerfile from here: https://github.com/Azure/azure-functions-docker/blob/dev/host/4/bullseye/amd64/node/node18/node18-core-tools.Dockerfile and since we use docker-compose  to bring up an Azurite and SQL service it was a really minor fix, but you definitely oughta get this updated!

**Comment by timtucker-dte** [comment]
Just started playing around.

The ability to use nested directories & define routes via code is a game changer.

In just a little bit of playing around, I'm able to pass __dirname from my index.ts files to utility functions to grab both the route and the method for a function based on following directory conventions.

Which leads to things like:
/functions/api/foo/{bar}/[GET]

Translating to:
Route:
foo/{bar}

Method:
GET

**Comment by charlesanim** [comment]
Hi, I just started playing with the v4 model to migrate our v3 functions. Our v3 functions is in internal production env, so i deployed changes leveraging v4 model and for unknown reasons the functions are not showing up anymore in the web portal. Any ideas? I'm not seeing any errors in logger. I don't see y'all document potential issues folks may run into and how to deal with it anywhere. My functions are written in Typescript, Env is Node 18 running on Linux. Any help will be appreciated.


i have my package.json like

`"main": "dist/src/functions/*.js",`

**Reply by ejizba** [reply]
Hi @charlesanim please see our troubleshooting guide here: https://learn.microsoft.com/en-us/azure/azure-functions/functions-node-troubleshoot?pivots=nodejs-model-v4

**Reply by Venipa** [reply]
the docs are as expected not to be of help, still not showing up as of today.

**Comment by Venipa** [comment]
i've created a new azure function service and tried to deploy v4 functions but it seems its not listing any functions,
locally it works.
tried using a bundler to ship the functions with a function.json but with no luck, no help from microsoft so far.

**Reply by ejizba** [reply]
Hi I noticed you filed an issue as well, I'll use that for followups: https://github.com/Azure/azure-functions-nodejs-library/issues/297

**Comment by sachinsom93** [comment]
We upgraded our Node.js functions from the **v3 model to the v4 model** around August 2024, using Node.js 18 at the time. As part of the migration, we also upgraded the underlying .NET framework to version 6.

Now, we’re planning to upgrade the Node.js version to 20 in the v4 model. Do we also need to upgrade the underlying .NET framework version as part of this update? Thanks!

**Comment by barddoo** [comment]
The Apollo GraphQL library integration that is recommended is not being maintained, so I forked it. Please check it out if you'd like to use GraphQL over Azure Functions. https://github.com/barddoo/apollo-server-integration-azure-functions
